lib/driver-utils.js

```diff
@@ -0,0 +1,210 @@
+  if (DRIVER_MAP[driverClass].driver) {
+    return DRIVER_MAP[driverClass].driver;
+  }
+  try {
+    log.debug(`Loading driver '${driverClass}' from '${packageName}'`);
+    const driver = require(packageName)[driverClass];
+
+    // cache the drivers
+    DRIVER_MAP[driverClass].driver = driver;
+
+    return driver;
+  } catch (err) {
+    log.errorAndThrow(`Unable to load driver '${driverClass}' from package '${packageName}': ${err.message}`);
+  }
+};
+
+function getDriverFromConstraints (caps, driverInfo) {
+  const getAlternativeDriver = function getAlternativeDriver (driverInfo) {
+    let alternativeDriver;
+    for (const constraint of (driverInfo.constraints || [])) {
+      if (constraint.type === 'capability') {
+        if (constraint.capability === 'platformVersion') {
+          const platformVersion = semver.valid(semver.coerce(caps.platformVersion));
+          if (platformVersion && semver.satisfies(platformVersion, constraint.condition)) {
+            if (constraint.message) {
+              log.warn(constraint.message);
+            }
+            if (constraint.alternativeDriver) {
+              alternativeDriver = constraint.alternativeDriver;
+            }
+          }
+        }
+      }
+    }
+    return alternativeDriver;
+  };
+
+  // drill down through constraints to get to the last driver that fits
+  let alternativeDriver = getAlternativeDriver(driverInfo);
+  while (alternativeDriver) {
+    const driver = getAlternativeDriver(DRIVER_MAP[alternativeDriver]);
+    // finish if there is no alternative, or the alternative is the same as
+    // the one we already have (this happens if the constraint is meant to just
+    // log information)
+    if (!driver || driver === alternativeDriver) {
+      break;
+    }
+    alternativeDriver = driver;
+  }
+
+  if (alternativeDriver) {
+    const modifiedDriverClass = DRIVER_MAP[alternativeDriver].driverClass;
+    const modifiedPackageName = DRIVER_MAP[alternativeDriver].packageName;
+    if (modifiedDriverClass && modifiedPackageName) {
+      return [modifiedPackageName, modifiedDriverClass];
+    }
+  }
+  return [];
+}
+
+function getDriverFromAutomationName (caps) {
+  let packageName;
+  let driverClass;
+  if (_.isString(caps.automationName)) {
+    const name = caps.automationName.toLowerCase();
+    for (const info of _.values(DRIVER_MAP)) {
+      if (_.toLower(info.automationName) !== name) {
+        continue;
+      }
+
+      // this fits the automationName
+      driverClass = info.driverClass;
+      packageName = info.packageName;
+
+      // now check if there are constraints on this driver
+      const [modifiedPackageName, modifiedDriverClass] = getDriverFromConstraints(caps, info);
+      if (modifiedPackageName && modifiedDriverClass) {
+        packageName = modifiedPackageName;
+        driverClass = modifiedDriverClass;
+      }
+      // no need to go further
+      break;
```

Comment:
we can return right here and get rid of intermediate variables

Reply:
nice